content/posts/2020/2020-02-19-dask-clusters-overview.md

```diff
@@ -0,0 +1,337 @@
+
+### Dask Gateway
+
+[Dask Gateway](https://gateway.dask.org/) is a central service for managing Dask clusters. It provides a secure API which multiple users can communicate with to request Dask servers. It can spawn Dask clusters on Kubernetes, Yarn or batch systems.
+
+This tool is targeted at IT administrators who want to enable their users to create Dask clusters, but want to maintain some centralized control instead of each user creating their own thing. This can also be useful for tracking Dask usage and setting per user limits.
+
+```python
+from dask.distributed import Client
+from dask_gateway import GatewayCluster
+
+cluster = GatewayCluster()
+client = Client(cluster)
+```
+
+For each user the commands for creating and using a gateway cluster are the same. It is down to the administrator to setup and manage the gateway server and configure [authentication via kerberos or Jupyter Hub](https://gateway.dask.org/authentication.html#). They should also provide [configuration](https://gateway.dask.org/configuration-user.html) to their users so that Dask Gateway knows how to connect to the gateway server. In a large organisation or institution the IT department also likely provisions the machines that staff are using, and so should be able to drop configuration files onto users computers.
+
+### Local CUDA Cluster
+
+The last cluster manager I'm going to cover is `LocalCUDACluster` from the [Dask CUDA](https://github.com/rapidsai/dask-cuda) package.
+
+This is slightly different than the other cluster managers in that it is constructing a Dask cluster which is specifically optimised for a single piece of hardware. In this case it is targeting machines with multiple GPUs like an [NVIDIA DGX-2](https://www.nvidia.com/en-gb/data-center/dgx-2/).
+
+The cluster manager closely follows the `LocalCluster` in that is creates resources locally on the current machine, but instead of creating one worker per CPU core it creates one per GPU. It also changes some of the configuration defaults to ensure good performance of GPU workloads.
+
+```python
+from dask.distributed import Client
+from dask_cuda import LocalCUDACluster
+
+cluster = LocalCUDACluster()
+client = Client(cluster)
+```
+
+This package also has an alternative Dask worker bash command called `dask-cuda-worker` which also modified the defaults of the Dask worker to ensure it is optimised for GPU work.
+
+## Future
+
+Now that we have laid out the current state of the Dask distributed cluster ecosystem let's discuss where we could go next.
+
+As shown at the beginning a Dask cluster is a combination of scheduler, workers and clients which enable distributed execution of Python functions. Setting up your own cluster on your own machines is straight forward, but there is such a variety of ways to provision infrastructure that we now have a number of ways of automating this.
```

Comment:
"scheduler, workers and clients", is clients really plural? I've never used multiple clients, maybe is just something I'm not aware of, just pointing this out to double-check whether this is a typo or is really something that can be done.

Reply:
It is possible to connect multiple clients. I've seen it done in some use cases, but it's probably not the primary use case. So changing this I think makes sense.

I'm actually working on adding clients to the cluster map dashboard, mainly just so that people can see when they've accidentally created the client object multiple times 😆.

Reply:
```suggestion
As shown at the beginning a Dask cluster is a combination of scheduler, workers and client which enable distributed execution of Python functions. Setting up your own cluster on your own machines is straight forward, but there is such a variety of ways to provision infrastructure that we now have a number of ways of automating this.
```